### ISSUE TITLE
Mini RFC: wasm-pack should use progressive enhancement in its output

### ISSUE THREAD

**Issue by fitzgen** [rustwasm/wasm-pack#298]
## Summary

`wasm-pack` should default to plain text output and use progressive enhancement to add bells and whistles such as colors, bold, and progress bars when the functionality is available.

## Motivation

`wasm-pack` should work seamlessly with all terminals on all platforms, and when its output is piped to another process or a log file. Users should never see garbled characters or color codes.

## Details

By default, `wasm-pack` should emit only plain text output to `stdout`.

We should use [the `atty` crate](https://crates.io/crates/atty) and the `$TERM` environment variable to determine whether `stdout` is a fully featured terminal. If `atty::is(atty::Stream::Stdout)` is true and the `$TERM` environment variable is not set to `"dumb"`, then `wasm-pack` should also emit colors, bold styles, and show progress bars when it makes sense.

To enforce consistency with these rules across the code base, there should be one central place that manages whether colors are printed or progress bars are displayed. All other code should be able to assume that the colors are used and progress bars are displayed, and be none the wiser if that is not actually the case under the covers. Luckily, we already have most of the printing logic isolated into the `progressbar` module. The exception is some `format!`ed string messages used to construct some `Error` instances.

To move forward, we should create a new `terminal` module that encapsulates all terminal output, progressive enhancement, and feature detection.

1. For styling text, we should refactor all existing usage of `style(...).bold().dim()` etc to use our own `terminal` utility function(s) that only apply the `style(...)` when the terminal supports it.

2. For progress bars, we should have a utility in `terminal` that creates the progress bars for the caller. If `stdout` is not a tty or `$TERM` is `"dumb"`, then it should return a progress bar created with `indicatif::ProgressBar::hidden`, which is a no-op progress bar.

## Related Issues

* https://github.com/rustwasm/wasm-pack/issues/284

## References

* https://crates.io/crates/atty
* https://en.wikipedia.org/wiki/Computer_terminal#Dumb_terminals

**Comment by fitzgen** [comment]
There was some discussion of this at todays WG meeting. Consensus was that where possible, we should push this into upstream crates, but also generally do wahtever it takes to get the behavior described here at the end of the day.

**Comment by killercup** [comment]
Started some implementation work at <https://github.com/killercup/output-rs>. Very much WIP and subject to rapid change.

**Comment by alexcrichton** [comment]
I wanted to try to kick the tires on this again recently, and it looks like changes to fix smaller bugs are starting to be blocked on this sort of larger scale revamp of the output. I wanted to jot down some notes that I've personally noticed, as well as an idea of how to make progress.

Some sample output I just saw from `wasm-pack build` looked like:

```

  [1/9] Checking `rustc` version...
  [2/9] Checking crate configuration...
  [3/9] Adding WASM target...
  info: component 'rust-std' for target 'wasm32-unknown-unknown' is up to date
  [4/9] Compiling to WASM...
      Finished release [optimized] target(s) in 0.02s
  [5/9] Creating a pkg directory...
  [6/9] Writing a package.json...
   [INFO]: Optional fields missing from Cargo.toml: 'description', 'repository', and 'license'. These are not necessary, but recommended
  [7/9] Copying over your README...
  wasm-bindgen 0.2.27 (6dfbb4be8)
  [8/9] wasm-bindgen already installed...
  [9/9] Running WASM-bindgen...
  :-) Done in 0 seconds
| :-) Your wasm pkg is ready to publish at "/home/alex/code/wasm-pack-template/pkg".
```

Some notes I'd have on this output are:

* Compiler error messages and Cargo messages aren't colorized
  - This is due to two parts, one is that wasm-pack currently logs all output of
    subcommands to log files. Another is that the usage of indcatif requires
    redirecting all output.

* Lots of output quickly becomes boilerplate
  - for example "Checking `rustc` versoin..." gets printed on all invocations of
    `wasm-pack build`. Most of this informational output is relevant for maybe
    the first run to see what's going on, but it quickly just becomes a lot of
    boilerplate that keeps scrolling past.

* One-off messages from subprocesses aren't suppressed.
  - After running wasm-pack build I see messages like:

    ```
    [3/9] Adding WASM target...
    info: component 'rust-std' for target 'wasm32-unknown-unknown' is up to date
    [4/9] Compiling to WASM...
    ```

    or

    ```
    [7/9] Copying over your README...
    wasm-bindgen 0.2.27 (6dfbb4be8)
    [8/9] wasm-bindgen already installed...
    ```

    these stray messages between the steps are "boilerplate informational"
    sometimes (like the rustup command) or sort of confusing and random (like
    the wasm-bindgen version getting printed)

* Warnings are unconditionally printed and can't be suppressed
  - Even for one-off projects informational warnings about missing `description`
    and such are printed, and there's not way to squelch the warning. This means
    that for some projects the warnings are always and unconditionally printed
    to the console and you've just got to get used to them being printed (which
    reduces the usefulness of the warning).

* Two "Done in XX seconds" messages are printed
  - one comes from Cargo and one comes from `wasm-pack`, and it's sort of
    confusing to see that because it looks like `wasm-pack` is finishing twice

* Indicatif output always seems somewhat buggy for me
  - For example my builds almost always end with:

    ```
    | :-) Your wasm pkg is ready to publish at "/home/alex/code/wasm-pack-template/pkg".
    ```

    I'm not sure why the spinner of `|` always shows up at the beginning of the
    line.

* `wasm-pack` produces practically zero output when output is redirected.
  - If you direct wasm-pack's output to a log file, all I see is an `[INFO]`
    message for a missing `description` and such, zero other progress
    indicators.

---

My personal opinion on what should be done to solve these issues is:

1. Don't assume that a number of steps is known, don't print every step.

2. Only print information if it's a warning, error, or informational about a
   step that's going to take a long time.
   * For example there's no need to print "Checking `rustc` version ..."
   * There's also no need to print "Adding WASM target..." unless it's not
     actually added already. `wasm-pack` could detect it's not installed and
     only print that if rustup will actually download something.
   * Messages like "Compiling to WASM..." don't need to be printed because Cargo
     will take care of all informational messages

3. Never print warnings that unconditionally show up and can't be suppressed.
   Ideally the warning about `description` and such missing could be on a
   publication step rather than the test/build steps.

4. Don't use indicatif for the entire build. This requires that all output
   flows through indicatif which is too painful of a restriction to implement a
   reasonable solution to many of the above points. While a spinner could be
   available for some steps, it's not necessary to have in general:

   * Long-running downloads via `rustup` already have a progress indicator
   * Cargo now has a progress indicator
   * Downloads of tools can use indicatif on a one-off basis to provide a
     progress indicator for their downloads.
   * Almost nothing else is long enough running to need a progress indicator

5. Output of commands should not make their way to log files. Like with indcatif
   for the entire build, this is just too difficult a restriction to work to
   implement solutions to many of the above pain points. I think the rationale
   for this is to have good crash logs, which while admirable I think we should
   figure out how to get in a different fashion.

Overall what I'd like to see is something like the following for a first run:

```
$ wasm-pack build
 [*] Installing the Rust wasm target ...
# ... output of `rustup target add ...`
# raw output of `cargo build`, hooked up directly to the terminal
 [*] Downloading `wasm-bindgen vX.Y.Z` ...
 [*] Downloading `wasm-opt vX.Y.Z` ...
 :-) Your wasm pkg is ready to publish at "..."
```

wasm-pack's primary output here is indicating that tools are being downloaded
when necessary and an informational message when it's done. An incremental build
would look like:

```
$ wasm-pack build
# raw output of `cargo build`, hooked up directly to the terminal
 :-) Your wasm pkg is ready to publish at "..."
```

No extra informational messages are printed to the terminal, and it's largely
Cargo that's printing information here. `wasm-pack` does, however, indicate
where the output is located after `wasm-bindgen` has finished executing.

Some known problems with this strategy I can think of are:

* Cargo prints "Finished ..." which isn't actually correct, as `wasm-pack` has
  more work to do. Similarly its timer doesn't include things like the rustup
  download nor following tool downloads.

* If `wasm-bindgen` takes a long time then wasm-pack doesn't print much.

I think these issues are far more minor than the ones above, though, and are
easy to fix in isolation.

---

I'm curious what others think about this though! If there's general consensus I'm fine signing myself up for implementing all this.

**Comment by fitzgen** [comment]
This is a great write up, thanks @alexcrichton!

I would like to add that I think we should also be doing a better job of *testing* output than we are right now. We do a pretty good job of testing functionality, but have basically zero tests for output that we emit.

We do have https://github.com/rustwasm/wasm-pack/issues/18 on file to start doing this sort of thing, but it looks like that crate has been deprecated since that issue was filed and `assert_cmd` is the new hotness.

**Comment by alexcrichton** [comment]
True! I would be more than willing to set up a test harness as well for the output. Fixing all of these associated issues will also be required to write tests, because if you try to write a test today the output will always be empty!

**Comment by danwilhelm** [comment]
Just for reference for this issue, I am posting the wasm-pack output on Windows Console (containing many notdef glyphs). The latest Windows Console (and hence Powershell) [does not display](https://github.com/Microsoft/console/issues/190) [emoji glyphs](https://github.com/PowerShell/PowerShell/issues/6931) - although support is [coming soon](https://github.com/Microsoft/WSL/issues/75#issuecomment-391155429).

![image](https://user-images.githubusercontent.com/6137185/50934911-0318a800-141f-11e9-833c-415d6c99c9f9.png)

**Comment by alexcrichton** [comment]
I believe this has all since been implemented, so closing!

**Comment by richard-delorenzi** [comment]
What version was this fixed in? I just downloaded 0.10.3 and still see no way to remove colour.

I did notice that piping it triggers no colour (is this the only way to control it). 
I note that it outputs to stderr, but if we connect stdout to a pipe (colour is turned off). If we connect stderr to the pipe colour is left on).

This is a strange, and as far as I can tell undocumented behaviour.
